// engine/input.js
const keysDown = Object.create(null);
const keysPressed = Object.create(null);
const keysReleased = Object.create(null);

const handleKeyDown = function(e) {
  const code = e.code;
  if (!keysDown[code]) {
    keysPressed[code] = true;
  }
  keysDown[code] = true;
};

const handleKeyUp = function(e) {
  const code = e.code;
  if (keysDown[code]) {
    keysReleased[code] = true;
  }
  keysDown[code] = false;
};

export const keyDown = function(code) {
  return keysDown[code] || false;
};

export const keyPressed = function(code) {
  return keysPressed[code] || false;
};

export const keyReleased = function(code) {
  return keysReleased[code] || false;
};

export const init = function() {
  window.addEventListener('keydown', handleKeyDown);
  window.addEventListener('keyup', handleKeyUp);
};

export const destroy = function() {
  window.removeEventListener('keydown', handleKeyDown);
  window.removeEventListener('keyup', handleKeyUp);
};

export const postUpdate = function() {
  for (const code in keysPressed) {
    keysPressed[code] = false;
  }
  for (const code in keysReleased) {
    keysReleased[code] = false;
  }
};
