// engine/input.js
const keysDown = Object.create(null);
const keysPressed = Object.create(null);
const keysReleased = Object.create(null);

function handleKeyDown(e) {
  const code = e.code;
  if (!keysDown[code]) {
    keysPressed[code] = true;
  }
  keysDown[code] = true;
}

function handleKeyUp(e) {
  const code = e.code;
  if (keysDown[code]) {
    keysReleased[code] = true;
  }
  keysDown[code] = false;
}

export function keyDown(code) {
  return keysDown[code] || false;
}

export function keyPressed(code) {
  return keysPressed[code] || false;
}

export function keyReleased(code) {
  return keysReleased[code] || false;
}

export function init() {
  window.addEventListener('keydown', handleKeyDown);
  window.addEventListener('keyup', handleKeyUp);
}

export function destroy() {
  window.removeEventListener('keydown', handleKeyDown);
  window.removeEventListener('keyup', handleKeyUp);
}

export function postUpdate() {
  for (const code in keysPressed) {
    keysPressed[code] = false;
  }
  for (const code in keysReleased) {
    keysReleased[code] = false;
  }
}
